### PR TITLE
[DNM][RFC][WIP] Add automation to move RocksDB files between BlueFS devices.

### PR DESCRIPTION
**Pull request by aclamk**
This PR tries to combat spillover from fast db to slow db.
During compaction sometimes there is huge spike of database size, like in https://bugzilla.redhat.com/show_bug.cgi?id=1708711 .
While this state is only temporary, files left on db.slow will remain there.

The idea of this PR is to allow moving of RocksDB files between BlueFS devices, without any change noticeable by RocksDB.
This rebalancing has 3 modes, dependent on % free space on fast device:
- when fast device has a lot of free space, try moving all files to fast
- when fast device is reasonably full, keep all files tagged "fast" on fast device
- when fast device is almost full, move files tagged "slow" to slow device

This seems to be related to PR #28960 and its modifications.

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -42,6 +42,93 @@ static void slow_discard_cb(void *priv, void* priv2) {
...
+      if (dev_id != -1) {
+        FileRef file;
+        {
+          std::lock_guard l(bluefs->lock);
+          //bluefs->lock.lock()
+          for (auto &d : bluefs->dir_map) {
+            std::string dir = d.first;
+            for (auto &r : d.second->file_map) {
+              if (file_name == dir + "/" + r.first) {
+                file = r.second;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: stop both loops at this point?

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -3022,3 +3132,177 @@ bool BlueFS::wal_is_rotational()
...
+  std::unique_lock l(lock);
+  ceph_assert(target_device < alloc.size());
+  ceph_assert(alloc[target_device]);
+  uint64_t need = f->fnode.size;
+  int r = _allocate_without_fallback(target_device,
+                             need,
+                             &temp_extents);
+  if (r != 0) {
+    //skip copying file
+    alloc[target_device]->release(temp_extents);
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: no need to release - allocate_without_fallback does that for you...

File: src/os/bluestore/BlueFS.cc [hunk 3]
```diff
@@ -3022,3 +3132,177 @@ bool BlueFS::wal_is_rotational()
...
+    return nullptr;
+  }
+
+  AioMoveFileContext* c = new AioMoveFileContext(this);
+  c->source = f;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: IMO this and below lines are a good candidates for being in ctor.

File: src/os/bluestore/BlueFS.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueFS.cc]: Thanks.

File: src/os/bluestore/BlueFS.cc [hunk 4]
```diff
@@ -3022,3 +3161,268 @@ bool BlueFS::wal_is_rotational()
...
+}
+
+void BlueFS::file_move_finished(AioMoveFileContext* c)
+{
+  dout(10) << __func__ << " on " << c->source->fnode.ino << " result=" << c->result << dendl;
+  ceph_assert(move_context == c);
+  bool needsync = false;
+  if (c->result) {
+    lock.lock();
+    if (c->source->num_reading.try_lock()) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: I don't completely understand what's the handling here. What happens if somebody is reading from the file? You silently undo the move? What happens if file has been changed during the move? You don't hold exclusive lock over it while moving,do you?

File: src/os/bluestore/BlueFS.cc [hunk 5]
```diff
@@ -3133,16 +3126,106 @@ bool BlueFS::wal_is_rotational()
...
+  if (it == file_map.end()) {
+    dout(1) << __func__ << " no files" << dendl;
+    return FileRef{};
+  }
   
+  for (uint64_t cnt = 0; cnt < 1000; cnt++) {
+    bluefs_fnode_t& fnode = it->second->fnode;
+    if (fnode.prefer_bdev == wants_on &&
+        fnode.extents.size() != 0 &&
+        fnode.extents[0].bdev == is_on) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: it's not enough to check extens[0] - file might have multiple extents at different devices simultaneously.

File: src/os/bluestore/BlueFS.cc [hunk 6]
```diff
@@ -3133,16 +3126,106 @@ bool BlueFS::wal_is_rotational()
...
+      move_last_ino = it->first;
+      return it->second;
+    }
+    it++;
+    if (it == file_map.end())
+      it = file_map.begin();
+    if(it->first == move_last_ino)
+      break;
+  };
+  move_last_ino = it->first;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: IMO this enumeration algorithm might miss objects  from time to time, here is the case:
- iterator incremented at line 3211
- for loop ends
- move_last_ino set to it
- it set to find(move_last_ino) +1 on the next entry to the func
hence move_last_ino isn't checked.

File: src/os/bluestore/BlueFS.cc [hunk 7]
```diff
@@ -3022,3 +3132,177 @@ bool BlueFS::wal_is_rotational()
...
+void BlueFS::aio_read_chunk(AioMoveFileContext* c)
+{
+  dout(30) << "rebalance " << __func__ << dendl;
+  ceph_assert(c->read_bl.length() == 0);
+  uint64_t length = c->source_it->length - c->source_offset;
+  if (length > 10 * 1024 * 1024)
+    length = 10 * 1024 * 1024;
+
+  c->state = BlueFS::AioMoveFileContext::reading;
+  uint8_t dev = c->source_it->bdev;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: you should check if extent is already at target device here - no need to move then

File: src/os/bluestore/BlueFS.cc [hunk 8]
```diff
@@ -3022,3 +3161,268 @@ bool BlueFS::wal_is_rotational()
...
+   * %all_on_db
+   * B | move db_on_slow->db_on_db
+   * %no_slow_on_db
+   * C | move slow_on_db->slow_on_slow
+   * %100
+   *
+   * To prevent circular operation distance ( %all_on_db - %no_slow_on_db ) must be larger then largest file.
+   */
+  utime_t now = ceph_clock_now();
+  if (double(move_next_check - now) > 0) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: I can't find any increments of move_next_check which is wrong IMO

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -10583,6 +10583,9 @@ void BlueStore::_kv_sync_thread()
 	int r = _balance_bluefs_freespace();
 	ceph_assert(r >= 0);
       }
+      if (bluefs) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Suggest to consider another thread for this stuff - kv_sync_thread is a very hot path hence it's better to leave it as simple as possible.
MempoolThread is probably the next choice...  Which should be renamed sooner or later to better reflect its purpose

File: src/os/bluestore/BlueFS.cc [hunk 10]
```diff
@@ -3022,3 +3161,268 @@ bool BlueFS::wal_is_rotational()
...
+    return;
+  }
+
+  std::unique_lock l(lock);
+  if (move_context)
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: I can see that move_context is used as a boolean value so far - so may be better use atomic instead and do the check before the lock...

File: src/os/bluestore/BlueFS.cc (hunk 4 above)
Comment (aclamk) [reply src/os/bluestore/BlueFS.cc]: @ifed01 You exactly touched most critical part.
Skipping for moment the fact that check if file has changed is missing,
do you think that copying file in background and on final step either committing it, or dropping it is valid?

File: src/os/bluestore/BlueStore.cc (hunk 9 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: Thanks for advice.

File: src/os/bluestore/BlueFS.cc (hunk 4 above)
Comment (ifed01) [reply src/os/bluestore/BlueFS.cc]: @aclamk - I don't see any unsolvable issues with this approach, looks viable...

File: src/os/bluestore/BlueStore.cc (hunk 9 above)
Comment (markhpc) [reply src/os/bluestore/BlueStore.cc]: Yes, MempoolThread is looking more like PeriodicBackgroundTaskThread. ;)

**Pull request closed without merging**